types/mithril/index.d.ts

```diff
@@ -1,863 +1,268 @@
-		then<U, V>(success: SuccessCallback<T, U>, error: ErrorCallback<V>): Thennable<U | V>;
-		catch?(error: ErrorCallback<T>): Thennable<T>;
-		catch?<U>(error: ErrorCallback<U>): Thennable<T | U>;
+	// Vnode children types
+	export type Child = Vnode<any, any> | string | number | boolean | null | undefined;
+	export interface ChildArray extends Array<Children> { }
+	export type Children = Child | ChildArray;
+	export type ChildArrayOrPrimitive = ChildArray | string | number | boolean;
+
+	/** Virtual DOM nodes, or vnodes, are Javascript objects that represent an element (or parts of the DOM). */
+	export interface Vnode<Attrs, State extends Lifecycle<Attrs, State>> {
+		/** The nodeName of a DOM element. It may also be the string [ if a vnode is a fragment, # if it's a text vnode, or < if it's a trusted HTML vnode. Additionally, it may be a component. */
+		tag: string | Component<Attrs, State>;
+		/** A hashmap of DOM attributes, events, properties and lifecycle methods. */
+		attrs: Attrs;
+		/** An object that is persisted between redraws. In component vnodes, state is a shallow clone of the component object. */
+		state: State;
+		/** The value used to map a DOM element to its respective item in an array of data. */
+		key?: string | number;
+		/** In most vnode types, the children property is an array of vnodes. For text and trusted HTML vnodes, The children property is either a string, a number or a boolean. */
+		children?: ChildArrayOrPrimitive;
+		/** This is used instead of children if a vnode contains a text node as its only child. This is done for performance reasons. Component vnodes never use the text property even if they have a text node as their only child. */
+		text?: string | number | boolean;
 	}
 
-	/**
-	* This represents a Mithril promise object.
-	*/
-	interface Promise<T> extends Thennable<T>, Property<T | Promise<T>> {
-		/**
-		* Chain this promise with a simple success callback, propogating
-		* rejections.
-		*
-		* @param success The callback to call when the promise is resolved.
-		* @return The chained promise.
-		*/
-		then<U>(success: SuccessCallback<T, U>): Promise<U>;
-
-		/**
-		* Chain this promise with a success callback and error callback, without
-		* propogating rejections.
-		*
-		* @param success The callback to call when the promise is resolved.
-		* @param error The callback to call when the promise is rejected.
-		* @return The chained promise.
-		*/
-		then<U, V>(success: SuccessCallback<T, U>, error: ErrorCallback<V>): Promise<U | V>;
-
-		/**
-		* Chain this promise with a single error callback, without propogating
-		* rejections.
-		*
-		* @param error The callback to call when the promise is rejected.
-		* @return The chained promise.
-		*/
-		catch<U>(error: ErrorCallback<U>): Promise<T | U>;
+	// In some lifecycle methods, Vnode will have a dom property
+	// and possibly a domSize property.
+	export interface VnodeDOM<Attrs, State> extends Vnode<Attrs, State> {
```

Comment:
You need to declare the generic `State` as `State extends Lifecycle<Attrs, State>`, due to the similar `Vnode` constraint above.